### PR TITLE
feat: implement user detail service

### PR DESCRIPTION
**Pull request by miamai**
**Main**
- [x] [951](https://www.notion.so/GET-api-auth-user-id-b060f66ebd0f4b8da4681cb9a00c7c16) GET user detail
- [x] [952](https://www.notion.so/PATCH-api-auth-user-id-de6b156f127a4f6e9e7f582d8044188f) PATCH user detail
- [x] config header token + 401 error

**Minor**
- [x] refactor type models

### CODE REVIEW

File: src/service/user.service.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import User from '../model/user.model';
+import { UserSchemaModel } from '../type/user.type';
+import { CoreError } from '../util/error-handler';
+
+export const get = async (userId: string) => {
+  const existingUser = await User.findById(userId);
```
Comment (cedarwud) [inline src/service/user.service.ts]: const existingUser = await User.findById(userId).select('-password');

**Comment by miamai** [comment]
Screenshot 2024-05-11 at 12 26 10 PM

當ObjectId不是 mongodb的數字時(eg: not in 24 digits)會有 CastError，所以會直接報錯，還要再handle嗎？因為mongodb自己的error 有點難抓，不然就是再catch 一個mongodb的error msg

**Comment by fufuShih** [comment]
看是不是要再sevice上的methods都先檢查userId是否合法，可能是像這樣:
```ts
export const get = async (userId: string) => {
  try {
    if (!userId.match(/^[0-9a-fA-F]{24}$/))
      throw new CoreError('Invalid userId');

    const existingUser = await User.findById(userId).select('-password');
    if (!existingUser) throw new CoreError('User not found');

    return existingUser;
  } catch (error) {
    console.log(error.message);
    throw new CoreError('User not fould');
  }
};
```

或者是創建一個新的middleware去檢查userId是否合法。
image

剛看到mongoose似乎有提供check。[Ref](https://github.com/Automattic/mongoose/issues/1959)
```
// Requiring ObjectId from mongoose npm package
const ObjectId = require('mongoose').Types.ObjectId;
 
// Validator function
function isValidObjectId(id){
     
    if(ObjectId.isValid(id)){
        if((String)(new ObjectId(id)) === id)
            return true;        
        return false;
    }
    return false;
}
```

File: src/model/user.model.ts [hunk 2]
```diff
@@ -43,6 +37,34 @@ const UserSchema = new Schema<IUser, UserModel, IUserMethods>(
...
+      type: Schema.Types.String,
+      enum: Gender,
+    },
+    age: {
+      type: Schema.Types.Number,
+    },
+    introduction: {
+      type: Schema.Types.String,
+    },
+    phoneAreaCode: {
```
Comment (cedarwud) [inline src/model/user.model.ts:56]: phoneAreaCode and phoneNumber should be string,otherwise the first '0' will disappear

Comment (miamai) [reply src/model/user.model.ts:56]: nice, add a validator for checking number as well

**Comment by miamai** [comment]
> 看是不是要再sevice上的methods都先檢查userId是否合法，可能是像這樣:
> 
> ```ts
> export const get = async (userId: string) => {
>   try {
>     if (!userId.match(/^[0-9a-fA-F]{24}$/))
>       throw new CoreError('Invalid userId');
> 
>     const existingUser = await User.findById(userId).select('-password');
>     if (!existingUser) throw new CoreError('User not found');
> 
>     return existingUser;
>   } catch (error) {
>     console.log(error.message);
>     throw new CoreError('User not fould');
>   }
> };
> ```
> 
> 或者是創建一個新的middleware去檢查userId是否合法。 image
> 
> 剛看到mongoose似乎有提供check。[Ref](https://github.com/Automattic/mongoose/issues/1959)
> 
> ```
> // Requiring ObjectId from mongoose npm package
> const ObjectId = require('mongoose').Types.ObjectId;
>  
> // Validator function
> function isValidObjectId(id){
>      
>     if(ObjectId.isValid(id)){
>         if((String)(new ObjectId(id)) === id)
>             return true;        
>         return false;
>     }
>     return false;
> }
> ```

found out had build-in validate function in `mongoose`

**Review by fufuShih — APPROVED** [review]

**Pull request merged**